### PR TITLE
[BEAM-12153] Implement GroupByKey with CombinePerKey with Concatenate CombineFn 

### PR DESCRIPTION
**Pull request by echauchot**
As flink runner does in GBK translation.

It is to avoid OOM due to the materialization of the collection in current GBK impl (see details in the ticket)

R: @ibzib 
CC: @aromanenko-dev 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

`ValidatesRunner` compliance status (on master branch)
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_ULR/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2_Streaming/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Samza/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Examples testing status on various runners
--------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Lang</th>
      <th>ULR</th>
      <th>Dataflow</th>
      <th>Flink</th>
      <th>Samza</th>
      <th>Spark</th>
      <th>Twister2</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Go</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Java</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Examples_Dataflow_Java11_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java_Examples_Dataflow_V2/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>Python</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
    <tr>
      <td>XLang</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

Post-Commit SDK/Transform Integration Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>Go</th>
      <th>Java</th>
      <th>Python</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
  </tbody>
</table>

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

<table>
  <thead>
    <tr>
      <th>---</th>
      <th>Java</th>
      <th>Python</th>
      <th>Go</th>
      <th>Website</th>
      <th>Whitespace</th>
      <th>Typescript</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td>Non-portable</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/">
          Build Status
        </a><br>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
    </tr>
    <tr>
      <td>Portable</td>
      <td>---</td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>
        <a href="https://ci-beam.apache.org/job/beam_PreCommit_GoPortable_Cron/lastCompletedBuild/">
          Build Status
        </a>
      </td>
      <td>---</td>
      <td>---</td>
      <td>---</td>
    </tr>
  </tbody>
</table>

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by echauchot** [comment]
Run Spark StructuredStreaming ValidatesRunner

File: runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java [hunk 1]
```diff
@@ -17,74 +17,143 @@
...
-        PTransform<PCollection<KV<K, V>>, PCollection<KV<K, Iterable<V>>>>> {
+  implements TransformTranslator<
+  PTransform<PCollection<KV<K, V>>, PCollection<KV<K, Iterable<V>>>>> {
+  /**
+   * Combiner that combines {@code T}s into a single {@code List<T>} containing all inputs.
+   *
+   * <p>For internal use to translate {@link GroupByKey}. For a large {@link PCollection} this is
+   * expected to crash!
+   *
+   * <p>This is copied from the dataflow runner code.
```
Comment (ibzib) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java:53]: It looks like there are already several implementations of this function. We should define it in a common place so we can reuse it. Possibly in the runner libraries, though it looks like Python even defines the equivalent function publicly: https://beam.apache.org/releases/pydoc/2.31.0/apache_beam.transforms.combiners.html#apache_beam.transforms.combiners.ToList

Comment (echauchot) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java:53]: Sure, I was thinking the same thing but I forgot to do the extraction. Indeed it is in DataFlow, Flink and Spark.

Comment (echauchot) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java:53]: well, just checked it is in more places than that I tracked the extraction in [this](https://issues.apache.org/jira/browse/BEAM-12727) ticket

**Comment by echauchot** [comment]
@ibzib thanks one again for prompt review !

**Comment by echauchot** [comment]
Run Flink ValidatesRunner

**Comment by echauchot** [comment]
Run Samza ValidatesRunner

**Comment by echauchot** [comment]
Run Spark ValidatesRunner

**Comment by echauchot** [comment]
Run Spark StructuredStreaming ValidatesRunner

**Comment by echauchot** [comment]
rebased on master

**Comment by echauchot** [comment]
Run Spark StructuredStreaming ValidatesRunner

**Comment by echauchot** [comment]
Run Spark StructuredStreaming ValidatesRunner

**Comment by echauchot** [comment]
Run Spark StructuredStreaming ValidatesRunner

File: runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java (hunk 1 above)
Comment (echauchot) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java]: I did not want to expose `Concatenate` transform in the sdk as it is internal and used only in the runners. So I put it into runner-core helpers package to be used for Spark, Flink and Samza. But dataflow does not depend on runner-core. I obviously did not want to add this dep so I left it duplicated in dataflow.

**Comment by echauchot** [comment]
Run Spark StructuredStreaming ValidatesRunner

**Comment by echauchot** [comment]
Run Java Precommit

**Comment by aaltay** [comment]
What is the next step on this PR?

**Comment by echauchot** [comment]
@aaltay next step is getting back from vacation, which just happened :smile: 
I'm gonna update the PR today.

**Comment by aaltay** [comment]
> @aaltay next step is getting back from vacation, which just happened 😄
> I'm gonna update the PR today.

Welcome back :)

**Comment by echauchot** [comment]
ES test split seems flaky (no split in my test)

**Comment by echauchot** [comment]
rebased on master

**Comment by echauchot** [comment]
> > @aaltay next step is getting back from vacation, which just happened smile
> > I'm gonna update the PR today.
> 
> Welcome back :)

thanks Ahmet !

**Comment by echauchot** [comment]
Run Spark StructuredStreaming ValidatesRunner

**Comment by echauchot** [comment]
@ibzib I addressed all your comments and validates runner tests pass. Still the java build is flaky with unrelated failures:
`org.apache.beam.runners.flink.FlinkSavepointTest.testSavepointRestoreLegacy
org.apache.beam.runners.dataflow.worker.StreamingDataflowWorkerTest.testHotKeyLoggingNotEnabled[1: [streamingEngine=true]]
org.apache.beam.runners.dataflow.worker.StreamingDataflowWorkerTest.testHotKeyLogging[1: [streamingEngine=true]]
org.apache.beam.sdk.io.elasticsearch.ElasticsearchIOTest.testSizes`

Can we merge the PR ?

**Comment by echauchot** [comment]
I did final commits cleanup, @aaltay if you're available, can we merge the PR ?

**Comment by echauchot** [comment]
Run Spark StructuredStreaming ValidatesRunner

File: runners/core-java/src/main/java/org/apache/beam/runners/core/helpers/package-info.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/** Provides helpers for Beam runner authors. */
+@DefaultAnnotation(NonNull.class)
+package org.apache.beam.runners.core.helpers;
```
Comment (ibzib) [inline runners/core-java/src/main/java/org/apache/beam/runners/core/helpers/package-info.java:21]: I think we can put `Concatenate` directly into `org.apache.beam.runners.core`. `org.apache.beam.runners.core` already serves the same purpose as the proposed `helpers` subpackage.

https://github.com/apache/beam/blob/3a7b8e757d0c220b5838ca6a0805ee463b24f85d/runners/core-java/src/main/java/org/apache/beam/runners/core/package-info.java#L19

**Review by ibzib — COMMENTED** [review]
Thanks Etienne, I like all the deleted lines :)

> @ibzib I addressed all your comments and validates runner tests pass. Still the java build is flaky with unrelated failures:

We don't want to merge PRs with test failures. If these tests are known flakes, you can propose sickbaying them.

File: runners/core-java/src/main/java/org/apache/beam/runners/core/helpers/package-info.java (hunk 2 above)
Comment (echauchot) [reply runners/core-java/src/main/java/org/apache/beam/runners/core/helpers/package-info.java:21]: agree, done

**Comment by echauchot** [comment]
> Thanks Etienne, I like all the deleted lines :)
> 
> > @ibzib I addressed all your comments and validates runner tests pass. Still the java build is flaky with unrelated failures:
> 
> We don't want to merge PRs with test failures. If these tests are known flakes, you can propose sickbaying them.

I agree, see my comment in related ticket: https://issues.apache.org/jira/browse/BEAM-5172?focusedCommentId=17409348&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-17409348
Sickbaying them.

**Comment by echauchot** [comment]
Run Spark StructuredStreaming ValidatesRunner

**Comment by echauchot** [comment]
Run Spark StructuredStreaming ValidatesRunner

**Comment by echauchot** [comment]
@ibzib I addressed all your latest comments and squashed. I also raised the priority of Elasticsearch falkes ticket that I ignored in this PR. Can we merge ?

**Comment by echauchot** [comment]
@aaltay , @ibzib can we merge this PR ?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java [hunk 3]
```diff
@@ -17,74 +17,95 @@
...
 import org.apache.beam.sdk.values.WindowingStrategy;
+import org.apache.spark.api.java.function.FlatMapFunction;
 import org.apache.spark.sql.Dataset;
 import org.apache.spark.sql.KeyValueGroupedDataset;
+import scala.Tuple2;
 
 class GroupByKeyTranslatorBatch<K, V>
-    implements TransformTranslator<
-        PTransform<PCollection<KV<K, V>>, PCollection<KV<K, Iterable<V>>>>> {
+  implements TransformTranslator<
```
Comment (ibzib) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java:43]: I am not sure why the indentation changed here. Do we not use Spotless on the Spark structured streaming project?

File: runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java [hunk 4]
```diff
@@ -17,74 +17,95 @@
...
     @SuppressWarnings("unchecked")
-    final PCollection<KV<K, V>> inputPCollection = (PCollection<KV<K, V>>) context.getInput();
-    Dataset<WindowedValue<KV<K, V>>> input = context.getDataset(inputPCollection);
-    WindowingStrategy<?, ?> windowingStrategy = inputPCollection.getWindowingStrategy();
-    KvCoder<K, V> kvCoder = (KvCoder<K, V>) inputPCollection.getCoder();
-    Coder<V> valueCoder = kvCoder.getValueCoder();
+    final PCollection<KV<K, V>> input = (PCollection<KV<K, V>>) context.getInput();
+    @SuppressWarnings("unchecked")
+    final PCollection<KV<K, List<V>>> output = (PCollection<KV<K, List<V>>>) context.getOutput();
+    @SuppressWarnings("unchecked")
```
Comment (ibzib) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java:55]: I don't think the suppression is necessary here. We're constructing a fresh object so it shouldn't count as an unchecked cast.

File: sdks/java/io/elasticsearch-tests/elasticsearch-tests-5/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java [hunk 5]
```diff
@@ -83,6 +84,7 @@ public void setup() throws IOException {
 
   @Rule public TestPipeline pipeline = TestPipeline.create();
 
+  @Ignore("Temporary ignore this test waiting for a fix because it is flaky")
```
[the remark this answers is not in the document]
Comment (echauchot) [reply sdks/java/io/elasticsearch-tests/elasticsearch-tests-5/src/test/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIOTest.java:87]: [the commit](https://github.com/apache/beam/pull/15267/commits/c0d4b37af8148ed8ccede80e314e7783d062620c) is already isolated with a separate jira but I can definitely move it to another PR.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java (hunk 3 above)
Comment (echauchot) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java:43]: we use spotless for this runner as well as it is in the same module as the other runner. Spotless seems happy with this indentation but I'll change it to conform to the other indentation used in the other places.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java (hunk 4 above)
Comment (echauchot) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java:55]: +1 thanks for pointing out !

**Comment by echauchot** [comment]
Run Spark StructuredStreaming ValidatesRunner

**Comment by echauchot** [comment]
@ibzib all addressed, PTAL

File: runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java (hunk 3 above)
Comment (ibzib) [reply runners/spark/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/GroupByKeyTranslatorBatch.java]: I don't have a problem with either way, I was just curious.

**Review by ibzib — APPROVED** [review]
LGTM!

**Pull request merged**